frontend/catalyst/device/decomposition.py

```diff
@@ -223,56 +226,125 @@ def measurements_from_counts(tape):
+                probs = _probs_from_counts(mapped_counts_outcome)
+                results_processed.append(_var_from_probs(eigvals=m.eigvals(), prob_vector=probs))
             elif isinstance(m, ProbabilityMP):
-                probs = _get_probs(mapped_counts_outcome)
+                probs = _probs_from_counts(mapped_counts_outcome)
                 results_processed.append(probs)
             elif isinstance(m, CountsMP):
                 results_processed.append(
                     tuple([states[0 : 2 ** len(m.wires)], mapped_counts_outcome])
                 )
+            else:
+                raise NotImplementedError(
+                    f"Measurement type {type(m)} is not implemented with measurements_from_counts"
+                )
         if len(tape.measurements) == 1:
             results_processed = results_processed[0]
         else:
             results_processed = tuple(results_processed)
         return results_processed
 
-    return [new_tape], postprocessing_counts_to_expval
+    return [new_tape], postprocessing_counts
+
+
+@transform
+@debug_logger
+def measurements_from_samples(tape, device_wires):
+    r"""Replace all measurements from a tape with sample measurements, and adds postprocessing
+    functions for each original measurement.
+
+    Args:
+        tape (QNode or QuantumTape or Callable): A quantum circuit.
+
+    Returns:
+        qnode (QNode) or quantum function (Callable) or tuple[List[QuantumTape], function]: The
+        transformed circuit as described in :func:`qml.transform <pennylane.transform>`.
+
+    .. note::
+
+        Counts are not supported.
+    """
+
+    new_operations, measured_wires = _diagonalize_measurements(tape, device_wires=device_wires)
+
+    new_tape = type(tape)(new_operations, [qml.sample(wires=measured_wires)], shots=tape.shots)
+
+    def postprocessing_samples(results):
+        """A processing function to get expecation values from samples."""
+        samples = results[0]
+        results_processed = []
+        for m in tape.measurements:
+            if isinstance(m, (ExpectationMP, VarianceMP, ProbabilityMP, SampleMP)):
+                if len(tape.shots.shot_vector) > 1:
+                    res = tuple(m.process_samples(s, measured_wires) for s in samples)
+                else:
+                    res = m.process_samples(samples, measured_wires)
+                results_processed.append(res)
+            else:
+                raise NotImplementedError(
+                    f"Measurement type {type(m)} is not implemented with measurements_from_samples"
+                )
+        if len(tape.measurements) == 1:
+            results_processed = results_processed[0]
+        else:
+            results_processed = tuple(results_processed)
+        return results_processed
+
+    return [new_tape], postprocessing_samples
+
+
+def _diagonalize_measurements(tape, device_wires):
```

Comment:
It kind of is, maybe it's silly to have this be separate. Getting the measurement wires is not part of the existing diagonalization functions, and they do "extra work" in creating diagonalized observables that we won't need anyway because we are taking samples directly with no observables in this transform. But that's not a lot of extra work, and we could combine them to have fewer things-that-do-almost-exactly-the-same-thing 🤷‍♀️ Maybe we should?

Reply:
Hmm I see, we'll I think it's fine as is as well :) Of course if you want to streamline in a follow up or in the future that's cool too 😄